### PR TITLE
Test restructure

### PR DESCRIPTION
**Pull request by matekelemen**
This is draft is still a WIP.

The primary task of this PR is to organize the existing tests and make them easily extendable. To this end, python's built-in ```unittest``` module is used. To run all tests that follow the ```unittest``` structure, execute [run_all_tests.py](https://github.com/mpentek/ParOptBeam/blob/test-documentation/test_scripts/run_all_tests.py), which recursively scans every script in ```test_scripts``` and collects them if:
- the file name begins with ```test_```
- the script has a class that inherits from ```unittest.TestCase``` or ```TestCase```
- said class has member functions beginning with ```test_```

Note that new tests can be added on the fly and will be discovered if they satisfy the requirements above.

By default, ```run_all_tests.py``` writes only the test results (which tests were run, passes, fails, errors) to the terminal. This output is piped into ```output/test_results.log``` as well, but the full captured terminal output is written to ```output/test_output.log```.

For the sake of convenience, I added an [analysis output parser](https://github.com/mpentek/ParOptBeam/blob/test-documentation/source/test_utils/parsed_results.py) that can also convert the results into a dictionary. This can be used to compare reference values and analysis output.

Apart from a handful of test scripts that already use ```unittest```, I only rewrote the [beam eigenvalue test](https://github.com/mpentek/ParOptBeam/blob/test-documentation/test_scripts/test_beam_eigenvalue_analytic.py) so far, which fails the ```fixed-pinned``` and ```pinned-pinned``` cases. I didn't dig too far in but it seems like this is a simple case of swapped values (```sway_y``` and ```sway_z```), at least for the ```fixed-pinned``` case. I'll mark the failing tests as expected failures until you decide what to do with them.

The rest of the tests that weren't rewritten to use ```unittest``` yet were modified such that they won't get executed unless directly called. I'll continue converting these scripts.

Once I converted them, I'll check how ```run_all_tests.py``` can be integrated into a CI on github. I'm confident I can finish the conversion by Thursday, but setting up a CI is new to me so I've no idea how long that's supposed to take.

### CODE REVIEW

**Comment by mpentek** [comment]
Sounds good! I saw that we actually have some `test_*` files which maybe do not make sense for the current state of the code. We can review these together.

The CI is not urgent for this week, but would be a nice touch to have.

By the way, when I launch `run_all_tests.py`, it complains that it cannot find the `from source [...]` that it cannot find the module. I did already have some issues with correct path and inits.

**Comment by matekelemen** [comment]
In a nuthsell, you can only run it from the root directory of the project, but that's true for any script importing from ```source```.

Here's what the project layout looks like currently:
- the important part of the repo is in ```source```, which is technically a module because it has an ```__init__``` script. ```source``` has submodules, each with their own ```__init__```s.
- tests are located outside the main module (```source```) in ```test_scripts```, which makes them standalone scripts that happen to be using ```source``` as a module.

As a result, ```source``` has to be discoverable as a module. There are two options to make this possible:
- the containing path of ```source``` has to be added to the global python path, so no matter where you execute a script from, ```source``` will always be found. If we were to go with this solution, I strongly recommend renaming the module from ```source``` to something more meaningful, like ```paroptbeam```. However, I don't recommend this option because appending the python path is tedius, different for each OS, and this module is not likely to be used from a wide array of locations anyway.
- alternatively, we can let users know that they must run scripts from the root directory of the project. Since the python interpreter searches the global python paths **and the current working directory of the terminal (or the IDE you're using)**, ```source``` is guaranteed to be found.

Personally, I'd go with the latter but still rename ```source``` to ```paroptbeam``` since it's a module and should have a descriptive name.

**Comment by mpentek** [comment]
Exactly, but I even have issues running it from root. Perhaps I am unlucky today with my setup.

**Comment by matekelemen** [comment]
It works for me know, but I'll try it with a clean install on another system.

**Comment by matekelemen** [comment]
I was wrong, python's search paths include the script's location and not working directory of the terminal (no idea why it worked on my main machine though). In this case, I recommend moving ```test_scripts``` into ```source```, and ```run_all_tests.py``` to root.

There's one caveat to this though: if you want to run an individual test script, you'll have to run it as a module:
```
python3 -m source.test_scripts.test_some_script.py
```

**Comment by matekelemen** [comment]
I converted the rests of the relevant tests, but none of them pass. The tests were defined in functions but those functions were never called in the scripts, so no tests were actually executed.

The most common error is a division by zero with the following trace:
- [StraightBeam::__init__](https://github.com/mpentek/ParOptBeam/blob/0743db8266277911d440ff7e99bb2a625f2bfc4c/source/model/structure_model.py#L171)
- [StraightBeam::identify_decoupled_eigenmodes](https://github.com/mpentek/ParOptBeam/blob/0743db8266277911d440ff7e99bb2a625f2bfc4c/source/model/structure_model.py#L605)
- [StraightBeam::decompose_and_quantify_eigenmodes](https://github.com/mpentek/ParOptBeam/blob/0743db8266277911d440ff7e99bb2a625f2bfc4c/source/model/structure_model.py#L565)

Then there's a missing ```LHS``` definition in the [BDF2 scheme](https://github.com/mpentek/ParOptBeam/blob/0743db8266277911d440ff7e99bb2a625f2bfc4c/source/solving_strategies/schemes/bdf2_scheme.py#L60).

Are these already addressed in master?

**Comment by mpentek** [comment]
> I was wrong, python's search paths include the script's location and not working directory of the terminal (no idea why it worked on my main machine though). In this case, I recommend moving `test_scripts` into `source`, and `run_all_tests.py` to root.
> 
> There's one caveat to this though: if you want to run an individual test script, you'll have to run it as a module:
> 
> ```
> python3 -m source.test_scripts.script_name_without_py_extension
> ```

I agree with this change.

**Comment by mpentek** [comment]
> I converted the rest of the relevant tests, but none of them pass. The tests were defined in functions but those functions were never called in the scripts, so no tests were actually executed.
> 
> The most common error is a division by zero with the following trace:
> 
> * [StraightBeam::**init**](https://github.com/mpentek/ParOptBeam/blob/0743db8266277911d440ff7e99bb2a625f2bfc4c/source/model/structure_model.py#L171)
> * [StraightBeam::identify_decoupled_eigenmodes](https://github.com/mpentek/ParOptBeam/blob/0743db8266277911d440ff7e99bb2a625f2bfc4c/source/model/structure_model.py#L605)
> * [StraightBeam::decompose_and_quantify_eigenmodes](https://github.com/mpentek/ParOptBeam/blob/0743db8266277911d440ff7e99bb2a625f2bfc4c/source/model/structure_model.py#L565)
> 
> Then there's a missing `LHS` definition in the [BDF2 scheme](https://github.com/mpentek/ParOptBeam/blob/0743db8266277911d440ff7e99bb2a625f2bfc4c/source/solving_strategies/schemes/bdf2_scheme.py#L60).
> 
> Are these already addressed in master?

No, these are not addressed yet. We should go through each test scrip to see if it up to date or just as a placeholder or downright wrong.

**Comment by matekelemen** [comment]
# Update

The primary goal of this PR is to provide a framework for testing whether the high-level behaviour changes while refactoring the repository. Keep in mind that this does **not** ensure the correctness of the results, but only helps locating new bugs.

The main idea is to store reference results in files, and overwrite them with current results when you want to create a snaphsot of the repo's behaviour. Generating such reference files can be done automatically, just make sure that you only overwrite the files you want to.

## Running Test Cases

All tests are implemented using python's builtin ```unittest``` module, so if you're familiar with that, you can skip this section.

You can run the main scripts located in the root directory just like you would execute any python script (for example ```python3 run_all_tests.py``` discovers and runs every test).

That said, if you want to run specific scripts or cases, you have to run them as modules because everyting within ```source``` is a module. For example, if you want to run all tests from *source/test_scripts/test_element.py*: ```python3 -m source.test_scripts.test_element```.

If you want to run a specific case in a script, say ```test_timoshenko_element``` in *source/test_scripts/test_element.py*:
```
python3 -m source.test_scripts.test_element -k test_timoshenko_element
```

Check out ```unittest```'s [documentation](https://docs.python.org/3/library/unittest.html) to learn more; it's useful.

## Generating References

If you want to make sure that all tests pass (assuming the tests don't throw exceptions), you can generate references anew and later test runs would compare their results to these newly generated ones. Keep in mind that this overwrites the existing reference files.

To do so, just set the ```--generate-references``` flag when running a test. Only the files used by the executed test will be overwritten. For example, if you want to overwrite the references of ```test_timoshenko_element``` in *source/test_scripts/test_element.py*:
```
python3 -m source.test_scripts.test_element -k test_timoshenko_element --generate-references
```

## Writing New Tests

A file comparison mechanism is overlayed on python's builtin ```unittest``` module to deal with automatic reference generation. This means that there is a custom [```TestCase```](https://github.com/mpentek/ParOptBeam/blob/test-documentation/source/test_utils/test_case.py) class you should derive your cases from (instead of ```unittest.TestCase```). This custom test case implements two relevant members:

- [```CompareToReferenceFile```](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_utils/test_case.py#L21) compares a numeric object (```numpy.ndarray```, or scalar like ```int``` or ```float```) to values stored in a reference file ([example](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_element.py#L23)).
- [```CompareFiles```](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_utils/test_case.py#L44) compares the contents of two files ([example](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_beam_eigenvalue_analytic.py#L133)).

Note, that any additional keyword arguments you pass to these functions will be forwarded to ```unittest.TestCase.assertAlmostEqual```, so you can control the behaviour of the comparison as you see fit (as an example, passing a ```delta``` keyword argument controls the absolute tolerance of numeric comparison).

When writing a new test, you can specify the path of reference files that don't yet exist, then run the test with the ```--generate-references``` flag set, and the reference files along with their parent direcories will be created for you. I recommend putting new reference files to *source/test_scripts/reference_output/<test_script_name>/<test_case_name>/<variable_name>.<file_extension>*.

## Issues

Some of these problems might already be mostly solved, but I'll list them anyway for completeness' sake:

- Division by zero errors in ```StraightBeam```'s constructor when calling [StraightBeam.decompose_and_quantify_eigenmodes](https://github.com/mpentek/ParOptBeam/blob/0743db8266277911d440ff7e99bb2a625f2bfc4c/source/model/structure_model.py#L479). This doesn't always happen but seems to be restricted to models that consist of a single element ([test_multiple_elements](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_multiple_elements.py), [test_residual_solver_static](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_residual_solver_static.py), [test_residual_solver_dynamic](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_residual_solver_dynamic.py))
- variable (```LHS```) not defined in the [BDF2 scheme](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/solving_strategies/schemes/bdf2_scheme.py#L60) affecting [test_schemes](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_schemes.py)
- some tests fail that have analytical references or references generated by Kratos ([test_crbeam_element_update_incremental](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_element_and_compare_to_kratos.py#L12), [test_schemes_analytic_sdof](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_schemes.py#L37)). I haven't overwritten because I guess *ParOptBeam* should produce these results.
- [test_structure_model_decompose_and_quantify_eigenmodes](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_structure_model_decompose_and_quantify_eigenmodes.py), [test_structure_model_eigenvalue_solve](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_structure_model_eigenvalue_solve.py) and [test_structure_model_identify_decoupled_eigenmodes](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_scripts/test_structure_model_identify_decoupled_eigenmodes.py) use mocking somewhat confusingly to me. I suspect it tries to circumvent the division by zero error by constructing ```StraightBeam```-like objects without calling its constructor, but unfortunately the mocking is not set up properly and no tests are executed at all (comparing empty lists). These scripts will have to be rewritten after the divisio by zero error is sorted out.
- generated references slightly vary depending on which machine generates them, possibly meaning that different machines yield different analyis results, but let's hope this is due to something I messed up in this PR (I'm checking it).

## Details

Two file types are supported:
- *.csv*: space-separated numeric values with a matching number of components in each column
- *.dat*: same as *.csv* but supports string values too (though parsing is much slower). Optinally, column names can be specified in a commented header line in the [expected format](https://github.com/mpentek/ParOptBeam/blob/ea67698f531f2e6a1696421f46f196c28bf7c843/source/test_utils/parsed_results.py#L336).

**Comment by mpentek** [comment]
Further notes on testing

Use the [analytical eigenform and values](https://github.com/mpentek/ParOptBeam/blob/8b4c6f8d9fce6baad1a02a7c9526ce39b40d5a82/source/test_scripts/analytical_solutions/analytical_eigen_beam_fixed_free.ipynb) to test from the [eigenvalue analysis](https://github.com/mpentek/ParOptBeam/blob/master/source/analysis/eigenvalue_analysis.py) (but some functionalities are tangled, as they are called inside the instance of [structure model](https://github.com/mpentek/ParOptBeam/blob/master/source/model/structure_model.py)):

- [the general solve](https://github.com/mpentek/ParOptBeam/blob/56c0d16d7ef4ebf635ee59290981ea9ace8d5c0f/source/analysis/eigenvalue_analysis.py#L45)
- [eigenvalue_solve](https://github.com/mpentek/ParOptBeam/blob/56c0d16d7ef4ebf635ee59290981ea9ace8d5c0f/source/model/structure_model.py#L617)
-   [decompose_and_quantify_eigenmodes](https://github.com/mpentek/ParOptBeam/blob/56c0d16d7ef4ebf635ee59290981ea9ace8d5c0f/source/model/structure_model.py#L479)
- [identify_decoupled_eigenmodes](https://github.com/mpentek/ParOptBeam/blob/56c0d16d7ef4ebf635ee59290981ea9ace8d5c0f/source/model/structure_model.py#L563)

The writing and plotting functions inside the eigenvalue analysis might give you a hint on how to receive data for eigenforms and frequencies. Ideally, you would have a loop over boundary conditions, taking into consideration fixed-free, free-fixed, fixed-fixed, pinned-pinned, possible fixed-pinned and pinned-fixed. See the definition of [bondary conditions](https://github.com/mpentek/ParOptBeam/blob/56c0d16d7ef4ebf635ee59290981ea9ace8d5c0f/source/auxiliary/global_definitions.py#L4). Potentially, you could have a JSON where these are swapped in the input, while defining an output only for the eigenvalue analysis block. With that you could continue using the file comparer, although it is maybe not the most elegant way of testing.

**Comment by matekelemen** [comment]
Test script: [test_beam_eigenvalue_analytic](https://github.com/mpentek/ParOptBeam/blob/d43f2781a188c54421818a5c2ab1522326cab30e/source/test_scripts/test_beam_eigenvalue_analytic.py)

## Eigenfrequencies

Eigenfrequencies are checked by [```BeamEigenvalueAnalyticalTest```](https://github.com/mpentek/ParOptBeam/blob/d43f2781a188c54421818a5c2ab1522326cab30e/source/test_scripts/test_beam_eigenvalue_analytic.py#L89) and they pass fine. It computes analytical eigenfrequencies for *y* and *z* bending modes, and checks whether the eigenfrequency list of ```StraightBeam``` has near matches.

Mode identification is not tested here, nor does this test catch wrong eigenfrequencies in the model.

Execute ```python3 -m source.test_scripts.test_beam_eigenvalue_analytic -k BeamEigenvalueAnalyticalTest``` from ```ParOptBeam``` root to run this specific test case.

## Eigenmode Identification

Identified eigenfrequencies are checked by [```BeamDecomposeEigenmodesAnalyticalTest```](https://github.com/mpentek/ParOptBeam/blob/d43f2781a188c54421818a5c2ab1522326cab30e/source/test_scripts/test_beam_eigenvalue_analytic.py#L185) which reveals some issues.

### ```pinned-pinned```

In the [```pinned-pinned```](https://github.com/mpentek/ParOptBeam/blob/d43f2781a188c54421818a5c2ab1522326cab30e/source/test_scripts/test_beam_eigenvalue_analytic.py#L203) case, the torsional rigid body mode gets misidentified as a bending mode in *z* direction, and pushes the indices of all other eigenfrequencies up by one. Furthermore, there's an extra eigenfrequency that isn't supposed to be in ```sway_z```. First 4 results for ```sway_z```:
- reference: ```[4.715861712387133, 18.863446849260722, 42.44275541148497, 75.45378739803449]```
- test: ```[nan  4.71573757  9.43028861 18.86146124]```

Execute ```python3 -m source.test_scripts.test_beam_eigenvalue_analytic -k BeamDecomposeEigenmodesAnalyticalTest.test_pinned_pinned``` from ```ParOptBeam``` root to run this specific test case.

### ```pinned-fixed```

There are 2 issues with the [```pinned-fixed```](https://github.com/mpentek/ParOptBeam/blob/d43f2781a188c54421818a5c2ab1522326cab30e/source/test_scripts/test_beam_eigenvalue_analytic.py#L203) case, which is surprising because the [```fixed-pinned```](https://github.com/mpentek/ParOptBeam/blob/d43f2781a188c54421818a5c2ab1522326cab30e/source/test_scripts/test_beam_eigenvalue_analytic.py#L197) case passes. Firstly, there are extra eigenfrequencies in ```sway_z```:
- reference: ```[7.367076030552972, 23.874045021953428, 49.811289323885525, 85.18025217996191]```
- test: ```[ 7.36684985 14.73165249 23.8713146  47.24816902]```

The other problem is that there's no ```torsional``` entry in ```StraightBeam::mode_identification_results``` (even though ```fixed-pinned``` has it). Note that you'll need to either fix the previous issues or comment the relevant assertions out to get to the point where this is checked (since the other failed assertions interrupt the test case). 

Execute ```python3 -m source.test_scripts.test_beam_eigenvalue_analytic -k BeamDecomposeEigenmodesAnalyticalTest.test_pinned_fixed``` from ```ParOptBeam``` root to run this specific test case.

## Running Test Cases in a Debugger

Since everything in ```ParOptBeam``` is within a python module (including test cases), initiating a debugging session is a bit tricky (at least in VSCode). You need to write a separate script which is not inside a module (eg.: in ```ParOptBeam``` root) and import the test case that you want to run, then execute this test script with the debugger. Example script to run the mentioned cases above:
```python
import source.test_scripts.test_beam_eigenvalue_analytic
from source.test_utils.test_case import TestMain

Case1 = source.test_scripts.test_beam_eigenvalue_analytic.BeamEigenvalueAnalyticalTest
Case2 = source.test_scripts.test_beam_eigenvalue_analytic.BeamDecomposeEigenmodesAnalyticalTest
TestMain()
```

To start a debugging session, you can write a debug config for VSCode or use [this](https://github.com/matekelemen/kript/blob/3aa594e63e9ba851a4e7d84913bb2e433e24da8a/scripts/run_debug) script that generates one for you and runs a debugging session on the script you pass to it (linux only):
```run_debug path_to_your_python_script.py -k BeamDecomposeEigenmodesAnalyticalTest.test_pinned_fixed```

**Comment by mpentek** [comment]
2 problems are probably occuring:

Issues with applying boundary conditions, thus the difference between fixed-pinned and pinned-fixed.

Issued with the mode identification, perhaps due to the nullification we use.

Please check if some of these do not appear when using 2D examples.

**Comment by matekelemen** [comment]
## Remaining Issues

### Eigenmode Identification

Run ```python -m source.test_scripts.test_beam_eigenvalue_analytic```.

- Computed eigenfrequencies and eigenmodes are correct, but some of them get identified in multiple directions. For example, in the fixed-fixed case, the first eigenmode in ```y``` direction gets identified both as ```sway_y``` and ```sway_z```.

### Dynamic Analysis

Run ```python -m source.test_scripts.test_schemes.TestSchemes```.

- The ```BDF2``` scheme always diverges, regardless of the time step size.
- Both nonlinear solvers (```ResidualBasedPicardSolver``` and ```ResidualBasedNewtonRaphsonSolver```) error out regardless of the setup. Maybe they're not designed to handle linear analysis?

**Comment by mpentek** [comment]
Testing `python -m source.test_scripts.test_beam_eigenvalue_analytic` I see some other issue or sources of confusion. In general, the `rel_participation` reference values need to be updated. I think the `GetModalProperties` of the `EulerBernoulli `beam are incorrect. The `modal_participation_factor` always has to be smaller than `1.0`.

`Fixed-Fixed` -> the observation above, and the fact that 3 eigenfrequencies are the same, but there are 2 differences. I am unsure how this happens. What I also see is, that in the `beam_model.decompose_and_quantify_eigenmodes() `we have entries that should not be there, namely every second. If you look at the `eff_modal_mass` or `rel_participation`, every second entry is numerically zero. These should have been nullified and not there as entries. For this test it means that
`sway_z `-> mode 1, 4, 9 and 13 should be identified, with the `rel_paticipation `factors 0.69, 0.13, 0.05, 0.029 (so 69%, 13%, 5%, 2.9%), which makes sense. This reasoning would make sense, but the `mode_id`-s and total count would not. For now counting up all the modes between `sway_z`, `sway_y `and torsion makes sense, with the above observation not anymore.

Curiously, when I run the run_test3deigen.py in the main folder, the results (marked as `eigenvalue_analysis_eigenmode_identification`) in the `output `folder for the `fixed-fixed` setting look good. I do not see the above observations.

```
======================================================================
FAIL: test_fixed_fixed (__main__.BeamDecomposeEigenmodesAnalyticalTest) [sway_z]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mate/Repositories/ParOptBeam/source/test_scripts/test_beam_eigenvalue_analytic.py", line 209, in runModel
    self.assertAlmostEqual(mode_info["rel_participation"], modal_participation_factor, delta = tolerance, msg = message_on_fail)
AssertionError: 0.6981665319146966 != 170159653.88417158 within 170159.65388417157 delta (170159653.18600506 difference) : 
test eigenfrequencies: [10.6899785  21.37685121 29.46468638 57.75443193]
reference eigenfrequencies: [10.690329210477406, 29.4683041220634, 57.769658516510525, 95.49618008692651]

[...]

----------------------------------------------------------------------
Ran 12 tests in 4.938s

FAILED (failures=12)
```

To summarize, there is an error in the analytic beam in how it calculates `eff_modal_mass` and `rel_participation`. There might still be a bug in the `decompose_and_identify_modes()`.

**Comment by mpentek** [comment]
> * Both nonlinear solvers (`ResidualBasedPicardSolver` and `ResidualBasedNewtonRaphsonSolver`) error out regardless of the setup. Maybe they're not designed to handle linear analysis?

These are not relevant for now, as the nonlinear solvers were not implemented properly.

**Comment by mpentek** [comment]
> ### Dynamic Analysis
> 
> Run `python -m source.test_scripts.test_schemes.TestSchemes`.
> 
>     * The `BDF2` scheme always diverges, regardless of the time step size.

With the current fix I receive with BDF2 the same results as with GenAlpha, both a bit different, than the reference, when I plot both. The difference might be due to scaling/zooming the results.

**Comment by mpentek** [comment]
Newest fix solves the modal mass participation and relative factor issue. Please check, I left with debug notes.

**Comment by mpentek** [comment]
WIth the latest commit the bug is fixed for the mode identification.

@matekelemen please go through the fixes, remove leftovers and merge into master. The testing branch we could now "freeze" be committing. Also, please add a brief "how-to" on how to use testing. Basically copying what we have here. I believe some tolerances might also need to be adjusted.

**Comment by matekelemen** [comment]
Ready to merge.

There's only one issue left though: the system tests ([TestGenericModels](https://github.com/mpentek/ParOptBeam/blob/50f34cd82aaabf89c44c09d3806bf2b31cc80c12/source/test_scripts/test_generic_models.py#L15)) yield slightly different results on different machines, which is pretty weird and a bit concerning.
